plugins/hls-class-plugin/src/Ide/Plugin/Class.hs

```diff
@@ -168,13 +217,17 @@ codeAction state plId (CodeActionParams _ _ docId _ context) = liftIO $ fmap (fr
         HAR {hieAst = hf} ->
           pure
             $ head . head
+#if MIN_VERSION_ghc(9,2,0)
+            $ pointCommand hf (fromJust (fromCurrentRange pmap range) ^. J.start & J.character +~ 1)
```

Comment:
Could you leave some comment on this difference? I mean `+~ 1` and `-~ 1`.

Reply:
I don't know, it just works after `+~ 1` and `-~ 1`.  The `-~ 1` is introduced by the first commit, I even don't know why it is required:(

As reference:
for file
```haskell
module T1 where

data X = X

instance Eq X where

```
The diag gives `Just (Range {_start = Position {_line = 4, _character = 9}, _end = Position {_line = 4, _character = 13}})` for both ghc-8.10.7 and ghc-9.2.1, then `pointCommand` run with the position `Position {_line = 4, _character = 10}` (ghc-9.2.1) and the position ` Position {_line = 4, _character =  8}`(ghc-8.10.7).

This means, on ghc-8.10.7, we search the position at the space before `Eq X`, and on ghc-9.2.1, we search the position at the character `q`.

Reply:
@July541 `-~ 1` is to make sure that the target contains `Eq` by cancelling https://github.com/haskell/haskell-language-server/blob/e8951c9152dbbd5a0a41b280ad76fc9e952164ab/ghcide/src/Development/IDE/Spans/AtPoint.hs#L414

I'm just not sure how `+~ 1` works

Reply:
@Ailrun Hello, I investigated the whole pipe:

1. (`+~ 1`) and (`-~ 1`) are totally different aspects, for (`-~ 1`), the [`TyThing`](https://hackage.haskell.org/package/ghc-9.2.1/docs/GHC-Types-TyThing.html#t:TyThing) we match is `AConLike` https://github.com/haskell/haskell-language-server/blob/ecec5396037cfd87da6f79091391d0ca2a194d0c/plugins/hls-class-plugin/src/Ide/Plugin/Class.hs#L231-L233. for (`+~ 1`) we match `ATyCon` https://github.com/haskell/haskell-language-server/blob/ef218f579bcea6dbdc571220c8388fcda6e6db87/plugins/hls-class-plugin/src/Ide/Plugin/Class.hs#L195. This means if we match `ATyCon`, both ghc8 and ghc9 are workable.
2. The real problem makes `codeAction` fail is the extra identifiers, https://github.com/haskell/haskell-language-server/blob/ecec5396037cfd87da6f79091391d0ca2a194d0c/plugins/hls-class-plugin/src/Ide/Plugin/Class.hs#L218-L224 Take `(Eq a) => Eq (X a)` as example, the  `pointCommand` result get `Eq` before, and in ghc9 it returns both `Eq` and `a`, which makes `head . head` fail.